server/src-lib/Hasura/RQL/Types/Common.hs

```diff
@@ -89,16 +92,31 @@ import           Hasura.SQL.Backend
 
 
 type family ScalarType (b :: Backend) where
-  ScalarType 'Postgres = PGScalarType
+  ScalarType 'Postgres = PG.PGScalarType
 
 type family ColumnType (b :: Backend) where
-  ColumnType 'Postgres = PGType
+  ColumnType 'Postgres = PG.PGType
 
 type family Column (b :: Backend) where
-  Column 'Postgres = PGCol
+  Column 'Postgres = PG.PGCol
 
 type family SQLExp (b :: Backend) where
-  SQLExp 'Postgres = S.SQLExp
+  SQLExp 'Postgres = PG.SQLExp
+
+class
+  ( Show (TableName b)
+  , Eq (TableName b)
+  , Lift (TableName b)
+  , NFData (TableName b)
+  , Cacheable (TableName b)
+  , Hashable (TableName b)
+  , Data (TableName b)
+  , Typeable b
+  ) => Representation (b :: Backend) where
+  type TableName b :: Type
+
+instance Representation 'Postgres where
```

Comment:
Why do we need the `Representation` typeclass? It only seems to define the `TableName` type family?

Reply:
Looks like Auke's comment above explains it:
> Whereas plain type families don't allow control over type class instances that are in scope, but type families associated to type classes do. This looks like an elegant and robust solution. We'll probably end up using this trick throughout our IR generalization work

Reply:
Every single type of the IR that wants to derive a class instance needs to explicitly list all its requirements, and **transitively the requirements of all its dependencies**: if a type has a field whose type has a field of type `TableName`, it will have to add a `Show (TableName b) =>` or `Eq (TableName b) =>` for each of its corresponding instances. The more types we generalize, the more unwieldy each new type declaration becomes... Using this new approach allows to list all the requirements on those types once, and then use a single `Representation b =>` at instance declaration time.

This PR introduces `Representation`, and subsequent PRs add to it. In the last commit of the pipeline it looks like this:
```
class
  ( Representable (Identifier b)
  , Representable (TableName b)
  , Representable (FunctionName b)
  , Representable (ConstraintName b)
  , Representable (BasicOrderType b)
  , Representable (NullsOrderType b)
  , Representable (Column b)
  , Representable (ScalarType b)
  , Representable (SQLExpression b)
  , Representable (SQLOperator b)
  , Ord (TableName b)
  , Ord (ScalarType b)
  , Lift (TableName b)
  , Lift (BasicOrderType b)
  , Lift (NullsOrderType b)
  , Data (TableName b)
  , Data (ScalarType b)
  , Data (SQLExpression b)
  , FromJSON (TableName b)
  , FromJSON (ScalarType b)
  , FromJSON (BasicOrderType b)
  , FromJSON (NullsOrderType b)
  , ToJSON (TableName b)
  , ToJSON (ScalarType b)
  , ToJSON (BasicOrderType b)
  , ToJSON (NullsOrderType b)
  , ToTxt (TableName b)
  , ToTxt (ScalarType b)
  , Typeable b
  ) => Representation (b :: Backend) where
  type Identifier     b :: Type
  type Alias          b :: Type
  type TableName      b :: Type
  type FunctionName   b :: Type
  type ConstraintName b :: Type
  type BasicOrderType b :: Type
  type NullsOrderType b :: Type
  type CountType      b :: Type
  type Column         b :: Type
  type ScalarType     b :: Type
  type SQLExpression  b :: Type
  type SQLOperator    b :: Type
```